File: python/day19.py
# S3 = {(), (0, 1, 2), (0, 2, 1), (0, 1), (1, 2), (0, 2)}
# Of those: (), (0, 1, 2), (0, 2, 1) are even and (0, 1), (1, 2), (2, 3) are odd
S3_WITH_SIGNS = [(1, (0, 1, 2)), (1, (1, 2, 0)), (1, (2, 0, 1)), (-1, (1, 0, 2)), (-1, (0, 2, 1)), (-1, (2, 1, 0))]
SIGNS = [-1, 1]

def l1_distance(a, b):
    return abs(a[0] - b[0]) + abs(a[1] - b[1]) + abs(a[2] - b[2])

def intersection_size(set1, set2):
    size = 0

    for x in set2:
        if x in set1:
            size += 1

    return size

def generate_rotations():
    # generate "permutation matrices with signs", to quote wikipedia:
    # "Take the set of all 3x3 permutation matrices and assign a + sign or a - 
    # sign to each of the three 1s. There are 6 permutations x 8 sign combinations = 48 matrices in 
    # total giving the full octahedral group. There are exactly 24 matrices with determinant = +1 
    # and these are the rotation matrices of the chiral octahedral group."
    # (https://en.wikipedia.org/wiki/Octahedral_symmetry)
    
    #...except we never actually construct the matrices since they don't
    # really help us (in matrix form) in any case

    for s, (i, j, k) in S3_WITH_SIGNS:
        for s1 in SIGNS:
            for s2 in SIGNS:
                for s3 in SIGNS:
                    # this is the determinant of the "signed permutation matrix"
                    # (the signs of the rows are (s1, s2, s3) and the rows are (e_i, e_j, e_k))
                    if s * s1 * s2 * s3 == -1:
                        continue

                    yield (s1, s2, s3, (i, j, k))

ROTATIONS = list(generate_rotations())

class BeaconSet:
    def __init__(self, beacons):
        self.data = beacons

        # Pre-compute a table of what distances each point in beacons
        # has to any other point in this set.
        self.distances = {}
        for p in self.data:
            self.distances[p] = set(l1_distance(p, q) for q in self.data)

    def __iter__(self):
        for x in self.data:
            yield x

    def rotate_and_translate(self, rot, d):
        s1, s2, s3, (i, j, k) = ROTATIONS[rot]
        d1, d2, d3 = d
        
        new_data = set()
        new_distances = {}
        for x in self.data:
            new_x = (s1 * x[i] - d1, s2 * x[j] - d2, s3 * x[k] - d3)

            new_data.add(new_x)
            new_distances[new_x] = self.distances[x]

        self.data = new_data
        self.distances = new_distances

def find_rotation_and_offset(set1: BeaconSet, set2: BeaconSet):
    # find some rotated version rset2 of set2 and an offset d such that
    # rset2 - d and set1 have an intersection of at least 12 elements.
    for p in set1:
        (p1, p2, p3) = p

        for q in set2:
            # This is a desperate attempt to limit the size of the search space.
            if intersection_size(set1.distances[p], set2.distances[q]) < 12:
                continue

            # p and q *might* be the same point.
            for rot in range(len(ROTATIONS)):
                s1, s2, s3, (i, j, k) = ROTATIONS[rot]

                (d1, d2, d3) = (s1 * q[i] - p1, s2 * q[j] - p2, s3 * q[k] - p3)

                m = 0
                for r in set2:
                    if (s1 * r[i] - d1, s2 * r[j] - d2, s3 * r[k] - d3) in set1:
                        m += 1

                if m >= 12:
                    return rot, (d1, d2, d3)

def solve():
    with open("problems/problem19") as f:
        lines = f.readlines()
        lines.append('\n')

    scanners = {}
    
    current = None
    current_point_set = set()
    for line in lines:
        line = line.strip()

        if line.startswith("---"):
            split = line.split(' ')
            current = int(split[2])
        elif not line:
            # Line is white space only, we have finished parsing the current scanners section
            scanners[current] = BeaconSet(current_point_set)
            current_point_set = set()
        else:
            x, y, z = [int(c) for c in line.split(',')]
            current_point_set.add((x, y, z))

    n = len(scanners)

    all_beacons = set()
    scanner_offsets = {0:(0, 0, 0)}

    while len(scanner_offsets) < n:
        # Pick any scanner for which we do not know an orientation, and try to find it
        # By comparing it to scanners for which we do know an orientation
        for t in scanners:
            if t in scanner_offsets:
                continue
                
            for s in scanner_offsets:
                res = find_rotation_and_offset(scanners[s], scanners[t])
                if res:
                    # (k1, k2, k3) is the offset of t relative to s
                    rot, offset = res

                    scanners[t].rotate_and_translate(rot, offset)
                    scanner_offsets[t] = offset

                    break


    # Assemble list of all beacons relative to scanner 0
    all_beacons = set()
    for s in scanners:
        all_beacons.update(scanners[s])

    print("Problem 1:", len(all_beacons))

    max_distance = 0

    for s, (p1, p2, p3) in scanner_offsets.items():
        for t, (q1, q2, q3) in scanner_offsets.items():
            if s == t:
                continue

            distance = abs(p1 - q1) + abs(p2 - q2) + abs(p3 - q3)

            if distance > max_distance:
                max_distance = distance

    print("Problem 2:", max_distance)

if __name__ == '__main__':
    solve()

File: src/day19.rs
use std::{
    ops::Sub,
};

use ahash::{AHashMap, AHashSet};
use lazy_static::lazy_static;

const INPUT: &str = include_str!("../problems/problem19");

// This is a re-implementation of the python version.
// See the Python version for more detailed comments.

#[derive(PartialEq, Eq, Hash, Clone, Debug, Copy)]
struct Point(i32, i32, i32);

impl Point {
    fn l1(&self) -> u32 {
        let &Point(p1, p2, p3) = self;
        p1.unsigned_abs() + p2.unsigned_abs() + p3.unsigned_abs()
    }

    fn component(&self, i: usize) -> i32 {
        match i {
            0 => self.0,
            1 => self.1,
            2 => self.2,
            _ => panic!("out of bounds"),
        }
    }

    fn rotate(&self, (s1, s2, s3, (i, j, k)): Rotation) -> Point {
        Point(
            s1 * self.component(i),
            s2 * self.component(j),
            s3 * self.component(k),
        )
    }
}

impl Sub for Point {
    type Output = Point;

    fn sub(self, Point(b1, b2, b3): Self) -> Self::Output {
        let Point(a1, a2, a3) = self;
        Point(a1 - b1, a2 - b2, a3 - b3)
    }
}

type Permutation = (usize, usize, usize);
type Rotation = (i32, i32, i32, Permutation);

const S3_WITH_SIGNS: [(i32, Permutation); 6] = [
    (1, (0, 1, 2)),
    (1, (1, 2, 0)),
    (1, (2, 0, 1)),
    (-1, (1, 0, 2)),
    (-1, (0, 2, 1)),
    (-1, (2, 1, 0)),
];

lazy_static! {
    static ref ROTATIONS: Vec<Rotation> = {
        let mut ret = Vec::new();

        for &(s, perm) in &S3_WITH_SIGNS {
            for s1 in [-1, 1] {
                for s2 in [-1, 1] {
                    for s3 in [-1, 1] {
                        if s * s1 * s2 * s3 == -1 {
                            continue;
                        }

                        ret.push((s1, s2, s3, perm))
                    }
                }
            }
        }

        ret
    };
}

struct BeaconSet {
    points_with_distances: AHashMap<Point, AHashSet<u32>>,
}

impl BeaconSet {
    fn from(set: AHashSet<Point>) -> Self {
        let mut points_with_distances: AHashMap<Point, AHashSet<u32>> = Default::default();

        for &x in &set {
            let x_distances = points_with_distances.entry(x).or_default();

            for &y in &set {
                x_distances.insert((x - y).l1());
            }
        }

        BeaconSet {
            points_with_distances,
        }
    }

    fn rotate_and_translate(&mut self, rot: Rotation, d: Point) {
        let n = self.points_with_distances.len();

        let old = std::mem::replace(&mut self.points_with_distances, AHashMap::with_capacity(n));

        for (x, distances) in old {
            let new_x = x.rotate(rot) - d;
            self.points_with_distances.insert(new_x, distances);
        }
    }
}

fn find_rotation_and_offset(set1: &BeaconSet, set2: &BeaconSet) -> Option<(Rotation, Point)> {
    for (&p, distances_p) in &set1.points_with_distances {
        for (&q, distances_q) in &set2.points_with_distances {
            if distances_p.intersection(distances_q).count() < 12 {
                continue;
            }

            for &rot in ROTATIONS.iter() {
                let d = q.rotate(rot) - p;

                let mut m = 0;
                for (r, _) in &set2.points_with_distances {
                    let t_and_r = r.rotate(rot)  - d;
                    if set1.points_with_distances.contains_key(&t_and_r) {
                        m += 1;
                    }
                }

                if m >= 12 {
                    return Some((rot, d));
                }
            }
        }
    }

    None
}

pub fn solve() -> crate::Result<()> {
    let lines = INPUT.lines().chain(std::iter::once("".into()));

    // Assume that the scanners are numbered 0,1,2,3... (without any holes)
    let mut scanners: Vec<BeaconSet> = Default::default();

    let mut current = None;
    let mut point_set: AHashSet<Point> = Default::default();

    for line in lines {
        let line = line.trim();

        if line.starts_with("---") {
            let mut split = line.split_ascii_whitespace();

            let nr = split.nth(2).ok_or(crate::Error::InvalidInput)?;
            current = Some(nr.parse::<usize>()?);
        } else if line.is_empty() {
            scanners.insert(
                current.ok_or(crate::Error::InvalidInput)?,
                BeaconSet::from(point_set),
            );
            point_set = Default::default();
        } else {
            let mut split = line.split(',');

            let x1 = split
                .next()
                .ok_or(crate::Error::InvalidInput)?
                .parse::<i32>()?;
            let x2 = split
                .next()
                .ok_or(crate::Error::InvalidInput)?
                .parse::<i32>()?;
            let x3 = split
                .next()
                .ok_or(crate::Error::InvalidInput)?
                .parse::<i32>()?;

            point_set.insert(Point(x1, x2, x3));
        }
    }

    let n = scanners.len();

    let mut scanner_offsets: AHashMap<usize, Point> = Default::default();
    scanner_offsets.insert(0, Point(0, 0, 0));

    while scanner_offsets.len() < scanners.len() {
        for t in 0..n {
            if scanner_offsets.contains_key(&t) {
                continue;
            }

            for &s in scanner_offsets.keys() {
                if let Some((rot, d)) = find_rotation_and_offset(&scanners[s], &scanners[t]) {
                    scanners[t].rotate_and_translate(rot, d);
                    scanner_offsets.insert(t, d);

                    break;
                }
            }
        }
    }

    // Assemble list of all beacons
    let mut all_beacons = AHashSet::new();
    for s in &scanners {
        for (&x, _) in &s.points_with_distances {
            all_beacons.insert(x);
        }
    }

    println!("Problem 1: {}", all_beacons.len());

    let mut max_distance = 0;
    for s in 0..n {
        for t in 0..n {
            if s == t {
                continue;
            }

            max_distance = std::cmp::max(
                (scanner_offsets[&s] - scanner_offsets[&t]).l1(),
                max_distance,
            )
        }
    }

    println!("Problem 2: {}", max_distance);

    Ok(())
}
